### DISCUSSION TITLE
.throttle doesn't fire a final event, potentially missing changes.

### DISCUSSION THREAD

**Discussion by samwillis** [alpinejs/alpine#2905] [7. Bugs (archived)]
If you are, for example, using `.throttle` for an auto search that's hitting the server repeatedly as the user enters text into a search box it will not fire a final event after they have finished typing. Chaining `.throttle.debounce` does not work as a solution.

I'm not sure what the "correct" behaviour is, I was however expecting it to fire a final event just as .debounce does after the timeout. The current docs doesn't really make it clear what to expect here, although by comparing it to debounce again I was expecting something closer to it.

(P.S. Alpine is awesome!)

**Comment by SimoTod** [comment]
The current implementation in alpine is that it only triggers the first event.
Let's pretend you have 3 keystroke close to each other, throttle will fire the callback on the first event and discard the last 2, debounce will discard the first 2 keystrokes and fire the callback on the last one.
It works like the one in this article: https://codeforgeek.com/throttle-function-javascript/
As far as I know, you can't use throttle and react at the same time. For your search box you probably want to use debounce rather than throttle.

**Reply by samwillis** [reply]
Thanks @SimoTod,

So I have spent some time reading up on different implementations of debounce/throttle this morning and it seems Alpine currently has a "trailing edge" debounce and "leading edge" throttle. This is a good summary of the differences: https://css-tricks.com/debouncing-throttling-explained-examples/

For my search box example it's a trailing edge throttle that's needed, debounce would stop it updating in real time as you type. It would only update when you stoped for more than the timeout.

I think there is probably a good argument for alpine to add a `.trailing` or `.leading` modifier to both `.debounce` and `.throttle`, so the user could. do something like:

```
@keydown.throttle.trailing.500ms="updateSearchbox()"
```

Some implementations (such as low dash) allow both trailing and leading on one function, I don't think alpine necessarily needs that.

https://lodash.com/docs/#debounce

https://lodash.com/docs/#throttle

(I'm not actually building something with a search box, I'm building a stateful/reactive Python/Django component toolkit, a little bit like Livewire. It's using Alpine on the front end. Going to show it off soon)

**Reply by SimoTod** [reply]
It might be a bit confusing to document (does the modifier needs to go before or after the time, or both) and complicate the implementation a bit but you can try to pitch it to Caleb. 

It may be easier to add it to Alpine.throttle initially as a third optional parameter and just use that in your callback `@keydown="Alpine.throttle(updateSearchbox, 500, true)"`.

Alternative, you can also implement a custom magic helper with the expected logic.

**Reply by samwillis** [reply]
Agreed, there are some complexities to it, however I think the current behaviour isn't necessarily what people will expect.

Adding it too `Alpine.throttle` would be good although using it in that way isn't currently documented at all. (It may be worth adding docs about using helpers on the global Alpine object?)

I assume to pitch it to Caleb I just convert this to an issue as explain the reasoning?

**Reply by SimoTod** [reply]
I don't know, probably easier if you send a PR with tests and explain the reasoning if it doesn't take too long. I don't think the issue tab is really used.

P.s. I would expect throttle to work as you described but apparently a lot of people on the Internet expect the current Alpine behaviour.

**Comment by rguca** [comment]
As a workaround you can use both:
```
let t = Alpine.throttle(() => this.updateItems(), 500);
let d = Alpine.debounce(() => this.updateItems(), 500);
input.addEventListener('input', () => t() & d());
```

**Reply by samwillis** [reply]
Thanks, that's a good suggestion. Maybe that's what changing the `.throttle.debounce` modifiers should actually do.

**Comment by pohlt** [comment]
It's all about expectations and if alpine users who are familiar with underscore.js or lodash.js use alpine's throttle mechanism, they are up for a surprise. 😮
Personally, I would like to see alpine's throttle behave exactly like _.throttle.
Here's a [small test page](https://codepen.io/pohlt/pen/poaGqNd) to compare both behaviors.

**Reply by SimoTod** [reply]
It would be a breaking change for other user, the only acceptable way would be to add a sensible modifier imo

**Reply by pohlt** [reply]
Well, there's always a 4.x, right? 😉 
How about `.final` because it makes sure that the final value is actually sent?

**Reply by 1aurabrown** [reply]
I would suggest using a `.trailing` modifier to mirror lodash's and underscore's terminology, given that most lodash's or underscore's throttle and debounce functions are the point of reference for most people, even if (unlike with lodash and underscore) `trailing` is false by default in alpine, to prevent introducing a breaking change for users.

**Comment by willcosgrove** [comment]
I just thought I would also add my current workaround into the mix in case anyone else is in need of a similar behavior. It appears to work to add multiple event listeners, so I do a throttled listener and a debounced listener.

```html
<input type="search" @input.debounce="$el.form.requestSubmit()" @input.throttle="$el.form.requestSubmit()" />
```

But I'll also just throw my two cents in and say that I expected throttle to have a trailing call. I'm having a difficult time thinking of an example of when that behavior would be undesirable.

**Comment by 1aurabrown** [comment]
I second adding the ability to have trailing functionality. As already pointed out, lodash does this well https://lodash.com/docs/4.17.15#throttle. My use case (like may other people's, probably), is throttling a scroll event.

In my case, the header of the website animates and with scroll position, so scroll position needs to be sampled at regular intervals (not simply debounced). The missing trailing fire on throttle means I don't get the user's final scroll position when they stop scrolling and the header doesn't look as it should.

One solution is to not use alpine's .throttle or .debounce, and to use something like lodash's trailing throttle method to create a throttled callback function myself to give to `@scroll`. I'll try @willcosgrove's solution above, too.
